vernac/auto_ind_decl.ml

```diff
@@ -699,18 +703,24 @@ let make_bl_scheme env handle mind =
   if not (Int.equal (Array.length mib.mind_packets) 1) then
     user_err
       (str "Automatic building of boolean->Leibniz lemmas not supported");
+
+  (* Setting universes *)
+  let auctx = Declareops.universes_context mib.mind_universes in
+  let u, uctx = UnivGen.fresh_instance_from auctx None in
+  let env = Environ.push_context_set uctx env in
+  let uctx = UState.update_sigma_univs (UState.of_context_set uctx) (Environ.universes env) in
```

Comment:
This should be 
```suggestion
  let uctx = UState.merge ~sideeff:false UState.univ_rigid (UState.from_env env) uctx in
```
Same for the other instances of update_sigma_univs 

The UState.of_context_set at line 183 is also suspiscious.

Reply:
I mostly tried to assemble code coming from other examples and I actually struggled on finding the right incantation. For instance, I found a bit weird that you have to build the env and the ustate in a crossed (and a bit convoluted) way. Naively, one would think that what we added to env from uctx does not need to be reconsidered and combined again with env to get a ustate. Why not a function to initialize at the same time the env and ustate from a fresh uctx?

Among other things which I find confusing, I already told it I think, it is the overloading of the name `uctx` for denoting either a `UState.t` (which is a rather complex structure) or a `ContextSet.t` (which is rather basic). Why not e.g. rather ustate for the former?

By the way, is using `update_sigma_univs` wrong or is it just not the recommended way to do?

> The UState.of_context_set at line 183 is also suspiscious.

I thought I had used the same `update_sigma_univs` incantation but this was maybe lost in a rebase.

Reply:
>   The UState.of_context_set at line 183 is also suspiscious.

Now I remember, the Boolean equality does not use tactics, nor sigma and this is why I did not have to change it.

Now, if the underlying ingredients (namely `andb`) were polymorphic and liable to add new constraints, that would probably be a problem, but `andb` has no universes and the scheme will not require more constraints than what is already present in the original intductive types, so I believe this is OK.

Reply:
I am not sure if adding the univs to the env is necessary/useful/a good idea.
No new update_sigma_univs uses should be added imo, it only exists for its current users.

Reply:
Also probably the scheme API should expect to get a (named_?)univ_entry, not a ustate.

Reply:
> I am not sure if adding the univs to the env is necessary/useful/a good idea.

Apparently, it works if I remove the line `let env = Environ.push_context_set uctx env in`. So, I apparently followed a wrong track, falling on missing universes which were actually missing in the ustate rather than in the env.

PR updated, using `UState.merge` and removing the `Environ.push_context_set`.

Reply:
> Also probably the scheme API should expect to get a (named_?)univ_entry, not a ustate.

Looks reasonable. Not sure I know how to do it though. `Ind_tables.define` minimizes and normalizes universes before building a univ_entry. So, would `auto_ind_decl.ml` have to do the minimize and nf_universes itself before returning a univ_entry?

It depends also on @ppedrot's projects. Maybe first waiting for knowing what is here orthogonal and what would be subsumed by his own PR.